src/dvclive/system_metrics.py

```diff
@@ -0,0 +1,87 @@
+import logging
+from typing import Dict, Union
+
+import psutil
+from statistics import mean
+from threading import Event, Thread
+
+from .utils import append_dict
+
+logger = logging.getLogger("dvclive")
+MEGABYTES_DIVIDER = 1024.0**2
+GIGABYTES_DIVIDER = 1024.0**3
+
+MINIMUM_CPU_USAGE_TO_BE_ACTIVE = 30
+
+
+class CPUMetricsCallback:
+    def __init__(
+        self,
+        duration: Union[int, float] = 5,
+        interval: Union[int, float] = 1.0,
+        plot: bool = True,
+    ):
+        self.duration = duration
+        self.interval = interval
+        self.plot = plot
+        self._no_plot_metrics = ["system/cpu/count", "system/cpu/ram_total_GB"]
+
+    def __call__(self, live):
+        self._live = live
+        self._shutdown_event = Event()
+        Thread(
+            target=self.monitoring_loop,
+        ).start()
+
+    def monitoring_loop(self):
+        while not self._shutdown_event.is_set():
+            self._cpu_metrics = {}
+            for _ in range(int(self.duration // self.interval)):
+                last_cpu_metrics = {}
+                try:
+                    last_cpu_metrics = get_cpus_metrics()
+                except psutil.Error:
+                    logger.exception("Failed to monitor CPU metrics:")
+                self._cpu_metrics = append_dict(self._cpu_metrics, last_cpu_metrics)
+                self._shutdown_event.wait(self.interval)
+                if self._shutdown_event.is_set():
+                    break
+            for metric_name, metric_values in self._cpu_metrics.items():
+                self._live.log_metric(
+                    metric_name,
+                    mean(metric_values),
+                    timestamp=True,
+                    plot=self.plot
+                    if metric_name not in self._no_plot_metrics
+                    else False,
+                )
+
+    def end(self):
+        self._shutdown_event.set()
+
+
+def get_cpus_metrics() -> Dict[str, Union[float, int]]:
+    ram_info = psutil.virtual_memory()
+    io_info = psutil.disk_io_counters()
+    nb_cpus = psutil.cpu_count()
+    cpus_percent = psutil.cpu_percent(percpu=True)
+    return {
+        "system/cpu/usage_avg_percent": mean(cpus_percent),
```

Comment:
how does it looks like across different products? I mean all these different names. Also, what do they show as X?

Reply:
### W&B
docs and naming: https://docs.wandb.ai/guides/app/features/system-metrics
W&B dashboard: https://wandb.ai/stacey/estuary?workspace=user-lavanyashukla

They show the timestamp in the X axis, but it is in hours and in respect of the beginning of the experiment. I really like their X axis to be honest. Names are spaced by " ". They do have the % usage per CPU core. It is a nice feature, but we will need to generate a complicated `dvc.yaml` configuration to display everything in one graph.

### labML
visual: https://github.com/labmlai/labml/blob/master/guides/hardware_monitoring.md
code: https://github.com/labmlai/labml/blob/6dc2a2ad3bd3a46c5dab9be80f3030c5ac8cb133/client/labml/internal/computer/monitor/scanner.py#L141
Their interface is way more simple than W&B. They use timestamp as the x-axis, but it is formatted nicely (day and hour, not a large int). They group metrics (like read/write on disk, memory available/remaining ...) which is nice and we should probably do the same. 

### Neptune
Dashboard: https://app.neptune.ai/o/showcase/org/onboarding-project/runs/details?viewId=98f66b32-2279-4b73-8210-863021c440ac&detailsTab=monitoring&shortId=IMG-155
doc: https://docs.neptune.ai/logging/system_metrics/

They are using very simple metrics, displayed by step (which I don't like). But it is quite easy to change the x-axis on their dashboard and they have relative timestamp and absolute timestamp as options. 


### Conclusion 
Imho, we should aim for timestamp as the default behavior. Ideally with a human display (rather than a large int). We should gather metrics together in one graph (like all the CPU cores ...). This should avoid creating weird "parallelization" metrics that are not well-defined.